### ISSUE TITLE
Stopping solver when total cost is below certain threshold

### ISSUE THREAD

**Issue by awa152** [ethz-adrl/control-toolbox#186]
I am encountering some interesting behavior, when solving my problem in an receding horizon MPC routine. 
When I update a new target pose, the first solution with GNMS takes 4 iterations and is shown below. 
For that I set printSummary and debugPrint to true. 

~~~
[MultipleShooting]: computing LQ Approximation from index 5 to N-1 took 1.547365 ms
[MultipleShooting]: Solving prepare stage of LQOC Problem
[MultipleShooting]: Prepare phase of LQOC problem took 0.002475 ms
[MultipleShooting]: prepareIteration() took 1.802402 ms
[MultipleShooting]: computing LQ approximation for first multiple-shooting interval took 0.184508 ms
[MultipleShooting]: Finish phase LQOC Problem
[MultipleShooting]: Finish solving LQOC problem took 7.557204 ms
[MultipleShooting]: Line search took 0.310682 ms
[MultipleShooting]: finishIteration() took 8.339397 ms
NLOC Summary of iteration 0
==============================
interm. cost:	1.37565964361
final cost:	0
total cost:	1.37565964361
total merit:	1.37565964361
tot. defect L1:	0.212041106208
tot. defect L2:	0.0973813909403
box cstr.err:	0
gen cstr.err:	0
total lx norm:	26.319324674
total lu norm:	14.9739679897
step-size:	1
                   ===========

[MultipleShooting]: computing LQ Approximation from index 5 to N-1 took 0.608909 ms
[MultipleShooting]: Solving prepare stage of LQOC Problem
[MultipleShooting]: Prepare phase of LQOC problem took 0.00051 ms
[MultipleShooting]: prepareIteration() took 0.635876 ms
[MultipleShooting]: computing LQ approximation for first multiple-shooting interval took 0.056361 ms
[MultipleShooting]: Finish phase LQOC Problem
[MultipleShooting]: Finish solving LQOC problem took 5.166026 ms
[MultipleShooting]: Line search took 0.329641 ms
[MultipleShooting]: finishIteration() took 5.602459 ms
NLOC Summary of iteration 1
==============================
interm. cost:	1.31085060474
final cost:	0
total cost:	1.31085060474
total merit:	1.31085060474
tot. defect L1:	0.0141449671345
tot. defect L2:	0.00663556822773
box cstr.err:	0
gen cstr.err:	0
total lx norm:	5.41752561809
total lu norm:	2.80196187147
step-size:	1
                   ===========

[MultipleShooting]: computing LQ Approximation from index 5 to N-1 took 1.163349 ms
[MultipleShooting]: Solving prepare stage of LQOC Problem
[MultipleShooting]: Prepare phase of LQOC problem took 0.0083 ms
[MultipleShooting]: prepareIteration() took 1.25171 ms
[MultipleShooting]: computing LQ approximation for first multiple-shooting interval took 0.132047 ms
[MultipleShooting]: Finish phase LQOC Problem
[MultipleShooting]: Finish solving LQOC problem took 9.610376 ms
[MultipleShooting]: Line search took 0.325102 ms
[MultipleShooting]: finishIteration() took 10.132516 ms
NLOC Summary of iteration 2
==============================
interm. cost:	1.30993248119
final cost:	0
total cost:	1.30993248119
total merit:	1.30993248119
tot. defect L1:	0.000102134243653
tot. defect L2:	5.92240828319e-05
box cstr.err:	0
gen cstr.err:	0
total lx norm:	0.626441226336
total lu norm:	0.467186996374
step-size:	1
                   ===========

[MultipleShooting]: computing LQ Approximation from index 5 to N-1 took 0.824793 ms
[MultipleShooting]: Solving prepare stage of LQOC Problem
[MultipleShooting]: Prepare phase of LQOC problem took 0.000609 ms
[MultipleShooting]: prepareIteration() took 0.858036 ms
[MultipleShooting]: computing LQ approximation for first multiple-shooting interval took 0.086972 ms
[MultipleShooting]: Finish phase LQOC Problem
[MultipleShooting]: Finish solving LQOC problem took 5.188592 ms
CONVERGED: Cost last iteration: 1.30993248119, current cost: 1.30992095926
CONVERGED: Cost improvement ratio was: 8.79581434371e-06x, which is lower than convergence criteria: 1e-05
[MultipleShooting]: Line search took 0.323568 ms
[MultipleShooting]: finishIteration() took 5.640796 ms
NLOC Summary of iteration 3
==============================
interm. cost:	1.30992095926
final cost:	0
total cost:	1.30992095926
total merit:	1.30992095926
tot. defect L1:	1.22957074048e-06
tot. defect L2:	5.56606391229e-07
box cstr.err:	0
gen cstr.err:	0
total lx norm:	0.0728548958297
total lu norm:	0.0599693717847
step-size:	1
                   ===========

~~~

When the new target is reached the iterations basically look like the one below, which take 2 iterations to converge.
The total cost is quite low, as the current position and the target position are nearly the same.

~~~
[MultipleShooting]: computing LQ Approximation from index 5 to N-1 took 0.918522 ms
[MultipleShooting]: Solving prepare stage of LQOC Problem
[MultipleShooting]: Prepare phase of LQOC problem took 0.001294 ms
[MultipleShooting]: prepareIteration() took 1.213108 ms
[MultipleShooting]: computing LQ approximation for first multiple-shooting interval took 0.110336 ms
[MultipleShooting]: Finish phase LQOC Problem
[MultipleShooting]: Finish solving LQOC problem took 3.195401 ms
[MultipleShooting]: Line search took 0.203796 ms
[MultipleShooting]: finishIteration() took 3.71993 ms
NLOC Summary of iteration 0
==============================
interm. cost:	2.5547500451e-05
final cost:	0
total cost:	2.5547500451e-05
total merit:	2.5547500451e-05
tot. defect L1:	2.51634709028e-10
tot. defect L2:	1.43073956603e-10
box cstr.err:	0
gen cstr.err:	0
total lx norm:	0.0014577312921
total lu norm:	0.00149407034827
step-size:	1
                   ===========

[MultipleShooting]: computing LQ Approximation from index 5 to N-1 took 0.478999 ms
[MultipleShooting]: Solving prepare stage of LQOC Problem
[MultipleShooting]: Prepare phase of LQOC problem took 0.001056 ms
[MultipleShooting]: prepareIteration() took 0.498655 ms
[MultipleShooting]: computing LQ approximation for first multiple-shooting interval took 0.082401 ms
[MultipleShooting]: Finish phase LQOC Problem
[MultipleShooting]: Finish solving LQOC problem took 1.668408 ms
CONVERGED: Cost last iteration: 2.5547500451e-05, current cost: 2.55475004509e-05
CONVERGED: Cost improvement ratio was: 4.98097471319e-12x, which is lower than convergence criteria: 1e-05
[MultipleShooting]: Line search took 0.259802 ms
[MultipleShooting]: finishIteration() took 2.050324 ms
NLOC Summary of iteration 1
==============================
interm. cost:	2.55475004509e-05
final cost:	0
total cost:	2.55475004509e-05
total merit:	2.55475004509e-05
tot. defect L1:	9.6662586789e-15
tot. defect L2:	5.57313032555e-15
box cstr.err:	0
gen cstr.err:	0
total lx norm:	3.23652730446e-07
total lu norm:	1.62208519741e-07
step-size:	1
                   ===========
~~~

After a while, when not updating the target pose, the total cost reaches values that are below 1e-20 and therefore the cost is basically zero. But from this point on with decreasing cost, the solver takes the maximum number of iterations which I allow him (in this case max_iterations 10) and does not converge. This is shown in the snippet below. It seems that the min_cost_improvement is higher than the set value of 1e-5. This seems like a strange behavior and I am not quite sure where this comes from. The extra iterations add computational burden, although the solution is already close to the optimal solution. Is there a way to tell the solver to stop, if the cost function is below a certain threshold?

~~~
[MultipleShooting]: computing LQ Approximation from index 5 to N-1 took 0.47438 ms
[MultipleShooting]: Solving prepare stage of LQOC Problem
[MultipleShooting]: Prepare phase of LQOC problem took 0.000645 ms
[MultipleShooting]: prepareIteration() took 0.63293 ms
[MultipleShooting]: computing LQ approximation for first multiple-shooting interval took 0.053868 ms
[MultipleShooting]: Finish phase LQOC Problem
[MultipleShooting]: Finish solving LQOC problem took 1.566991 ms
[MultipleShooting]: Line search took 0.293835 ms
[MultipleShooting]: finishIteration() took 2.093375 ms
NLOC Summary of iteration 0
==============================
interm. cost:	3.28868273545e-24
final cost:	0
total cost:	3.28868273545e-24
total merit:	3.28868273545e-24
tot. defect L1:	8.42533508239e-15
tot. defect L2:	4.26644419579e-15
box cstr.err:	0
gen cstr.err:	0
total lx norm:	1.10425255353e-12
total lu norm:	7.82944425813e-13
step-size:	1
                   ===========

[MultipleShooting]: computing LQ Approximation from index 5 to N-1 took 0.583684 ms
[MultipleShooting]: Solving prepare stage of LQOC Problem
[MultipleShooting]: Prepare phase of LQOC problem took 0.001324 ms
[MultipleShooting]: prepareIteration() took 0.619349 ms
[MultipleShooting]: computing LQ approximation for first multiple-shooting interval took 0.074126 ms
[MultipleShooting]: Finish phase LQOC Problem
[MultipleShooting]: Finish solving LQOC problem took 2.2091 ms
[MultipleShooting]: Line search took 0.232863 ms
[MultipleShooting]: finishIteration() took 2.568211 ms
NLOC Summary of iteration 1
==============================
interm. cost:	3.28881634585e-24
final cost:	0
total cost:	3.28881634585e-24
total merit:	3.28881634585e-24
tot. defect L1:	2.45810316546e-15
tot. defect L2:	1.66204568742e-15
box cstr.err:	0
gen cstr.err:	0
total lx norm:	3.25998057416e-14
total lu norm:	1.13660035377e-14
step-size:	1
                   ===========

[MultipleShooting]: computing LQ Approximation from index 5 to N-1 took 0.863663 ms
[MultipleShooting]: Solving prepare stage of LQOC Problem
[MultipleShooting]: Prepare phase of LQOC problem took 0.000267 ms
[MultipleShooting]: prepareIteration() took 0.925183 ms
[MultipleShooting]: computing LQ approximation for first multiple-shooting interval took 0.138291 ms
[MultipleShooting]: Finish phase LQOC Problem
[MultipleShooting]: Finish solving LQOC problem took 2.634974 ms
[MultipleShooting]: Line search took 0.168189 ms
[MultipleShooting]: finishIteration() took 2.979734 ms
NLOC Summary of iteration 2
==============================
interm. cost:	3.28852413276e-24
final cost:	0
total cost:	3.28852413276e-24
total merit:	3.28852413276e-24
tot. defect L1:	1.58293517183e-15
tot. defect L2:	1.2237391438e-15
box cstr.err:	0
gen cstr.err:	0
total lx norm:	2.41218172997e-14
total lu norm:	8.95038252024e-15
step-size:	1
                   ===========

[MultipleShooting]: computing LQ Approximation from index 5 to N-1 took 0.440116 ms
[MultipleShooting]: Solving prepare stage of LQOC Problem
[MultipleShooting]: Prepare phase of LQOC problem took 0.000263 ms
[MultipleShooting]: prepareIteration() took 0.453815 ms
[MultipleShooting]: computing LQ approximation for first multiple-shooting interval took 0.044917 ms
[MultipleShooting]: Finish phase LQOC Problem
[MultipleShooting]: Finish solving LQOC problem took 1.251118 ms
[MultipleShooting]: Line search took 0.222637 ms
[MultipleShooting]: finishIteration() took 1.557543 ms
NLOC Summary of iteration 3
==============================
interm. cost:	3.28870013945e-24
final cost:	0
total cost:	3.28870013945e-24
total merit:	3.28870013945e-24
tot. defect L1:	1.65340831304e-15
tot. defect L2:	1.3163268746e-15
box cstr.err:	0
gen cstr.err:	0
total lx norm:	2.38395660971e-14
total lu norm:	1.10918466669e-14
step-size:	1
                   ===========

[MultipleShooting]: computing LQ Approximation from index 5 to N-1 took 0.528867 ms
[MultipleShooting]: Solving prepare stage of LQOC Problem
[MultipleShooting]: Prepare phase of LQOC problem took 0.00027 ms
[MultipleShooting]: prepareIteration() took 0.552437 ms
[MultipleShooting]: computing LQ approximation for first multiple-shooting interval took 0.056111 ms
[MultipleShooting]: Finish phase LQOC Problem
[MultipleShooting]: Finish solving LQOC problem took 1.713249 ms
[MultipleShooting]: Line search took 0.200774 ms
[MultipleShooting]: finishIteration() took 2.023158 ms
NLOC Summary of iteration 4
==============================
interm. cost:	3.28851644819e-24
final cost:	0
total cost:	3.28851644819e-24
total merit:	3.28851644819e-24
tot. defect L1:	2.0554304786e-15
tot. defect L2:	1.47500058456e-15
box cstr.err:	0
gen cstr.err:	0
total lx norm:	2.28980907349e-14
total lu norm:	1.06372254528e-14
step-size:	1
                   ===========

[MultipleShooting]: computing LQ Approximation from index 5 to N-1 took 0.531926 ms
[MultipleShooting]: Solving prepare stage of LQOC Problem
[MultipleShooting]: Prepare phase of LQOC problem took 0.000224 ms
[MultipleShooting]: prepareIteration() took 0.550905 ms
[MultipleShooting]: computing LQ approximation for first multiple-shooting interval took 0.051623 ms
[MultipleShooting]: Finish phase LQOC Problem
[MultipleShooting]: Finish solving LQOC problem took 1.154027 ms
[MultipleShooting]: Line search took 0.190271 ms
[MultipleShooting]: finishIteration() took 1.425162 ms
NLOC Summary of iteration 5
==============================
interm. cost:	3.28871527512e-24
final cost:	0
total cost:	3.28871527512e-24
total merit:	3.28871527512e-24
tot. defect L1:	1.86114144929e-15
tot. defect L2:	1.32315381292e-15
box cstr.err:	0
gen cstr.err:	0
total lx norm:	2.24836443448e-14
total lu norm:	1.09094994722e-14
step-size:	1
                   ===========

[MultipleShooting]: computing LQ Approximation from index 5 to N-1 took 0.431078 ms
[MultipleShooting]: Solving prepare stage of LQOC Problem
[MultipleShooting]: Prepare phase of LQOC problem took 0.000272 ms
[MultipleShooting]: prepareIteration() took 0.447214 ms
[MultipleShooting]: computing LQ approximation for first multiple-shooting interval took 0.046213 ms
[MultipleShooting]: Finish phase LQOC Problem
[MultipleShooting]: Finish solving LQOC problem took 1.034877 ms
[MultipleShooting]: Line search took 0.148358 ms
[MultipleShooting]: finishIteration() took 1.251958 ms
NLOC Summary of iteration 6
==============================
interm. cost:	3.2885161854e-24
final cost:	0
total cost:	3.2885161854e-24
total merit:	3.2885161854e-24
tot. defect L1:	2.14650346109e-15
tot. defect L2:	1.42639357512e-15
box cstr.err:	0
gen cstr.err:	0
total lx norm:	2.27114004542e-14
total lu norm:	1.10804170671e-14
step-size:	1
                   ===========

[MultipleShooting]: computing LQ Approximation from index 5 to N-1 took 0.470987 ms
[MultipleShooting]: Solving prepare stage of LQOC Problem
[MultipleShooting]: Prepare phase of LQOC problem took 0.000709 ms
[MultipleShooting]: prepareIteration() took 0.48419 ms
[MultipleShooting]: computing LQ approximation for first multiple-shooting interval took 0.048406 ms
[MultipleShooting]: Finish phase LQOC Problem
[MultipleShooting]: Finish solving LQOC problem took 1.524463 ms
[MultipleShooting]: Line search took 0.186075 ms
[MultipleShooting]: finishIteration() took 1.787954 ms
NLOC Summary of iteration 7
==============================
interm. cost:	3.28872129738e-24
final cost:	0
total cost:	3.28872129738e-24
total merit:	3.28872129738e-24
tot. defect L1:	2.36833122558e-15
tot. defect L2:	1.58434773068e-15
box cstr.err:	0
gen cstr.err:	0
total lx norm:	2.40905876824e-14
total lu norm:	1.13845519168e-14
step-size:	1
                   ===========

[MultipleShooting]: computing LQ Approximation from index 5 to N-1 took 0.60261 ms
[MultipleShooting]: Solving prepare stage of LQOC Problem
[MultipleShooting]: Prepare phase of LQOC problem took 0.001238 ms
[MultipleShooting]: prepareIteration() took 0.644861 ms
[MultipleShooting]: computing LQ approximation for first multiple-shooting interval took 0.097242 ms
[MultipleShooting]: Finish phase LQOC Problem
[MultipleShooting]: Finish solving LQOC problem took 4.771079 ms
[MultipleShooting]: Line search took 0.362574 ms
[MultipleShooting]: finishIteration() took 5.308231 ms
NLOC Summary of iteration 8
==============================
interm. cost:	3.28855135816e-24
final cost:	0
total cost:	3.28855135816e-24
total merit:	3.28855135816e-24
tot. defect L1:	1.97823528392e-15
tot. defect L2:	1.52524870452e-15
box cstr.err:	0
gen cstr.err:	0
total lx norm:	2.26347477522e-14
total lu norm:	1.02842944755e-14
step-size:	1
                   ===========

[MultipleShooting]: computing LQ Approximation from index 5 to N-1 took 0.989535 ms
[MultipleShooting]: Solving prepare stage of LQOC Problem
[MultipleShooting]: Prepare phase of LQOC problem took 0.000959 ms
[MultipleShooting]: prepareIteration() took 1.047483 ms
[MultipleShooting]: computing LQ approximation for first multiple-shooting interval took 0.125472 ms
[MultipleShooting]: Finish phase LQOC Problem
[MultipleShooting]: Finish solving LQOC problem took 4.230373 ms
[MultipleShooting]: Line search took 0.217366 ms
[MultipleShooting]: finishIteration() took 4.630155 ms
NLOC Summary of iteration 9
==============================
interm. cost:	3.2887120875e-24
final cost:	0
total cost:	3.2887120875e-24
total merit:	3.2887120875e-24
tot. defect L1:	2.25644156138e-15
tot. defect L2:	1.67321335857e-15
box cstr.err:	0
gen cstr.err:	0
total lx norm:	2.46348231876e-14
total lu norm:	1.12360887914e-14
step-size:	1
                   ===========
~~~

I use HPIPM and my solver settings are:
~~~
solver_settings
{
    nlocp_algorithm GNMS
    integrator EulerCT
    useSensitivityIntegrator false
    discretization Forward_euler
    timeVaryingDiscretization false
    dt 0.04
    K_sim 1
    K_shot 5
    epsilon 0
    max_iterations 3
    fixedHessianCorrection false
    recordSmallestEigenvalue false
    min_cost_improvement 1e-5
    maxDefectSum 1e-5
    meritFunctionRho 0.0
    meritFunctionRhoConstraints 0.0
    nThreads 1
    nThreadsEigen 1
    printSummary true 
    debugPrint true
    ; logToMatlab 0   
    
    line_search
    {
        type NONE
        adaptive false
        maxIterations 10
        alpha_0 1.0
        n_alpha 0.5
        debugPrint true
    }
}
~~~

**Comment by markusgft** [comment]
Hi, my recommendation is to run only a single iteration in every mpc cycle!

**Comment by awa152** [comment]
Hi @markusgft,
thank you for the fast answer. If we don't consider the computational effort for a moment, wouldn't be a single iteration to little for the solver to come close to an optimal solution? 

Also I noticed that despite giving the MPC class a NLOptConSettings instance with a max_iterations count of greater than one, it always seems to make only a single iteration. At least it always prints out an Iteration count of 0 when I set printSummary to true. I don't know if this is intended like this. 
That's why I use NLOptConSolver directly and if needed the solver takes the amount of iterations set in max_iterations.
